icefoganalytics/internal-data-portal
# Part 2: Data Owner: Add Ability To Define Access Constraints

Relates to:
- [0001, Data Modeling, 2024-01-05](https://docs.google.com/document/d/1Mdbl_Cy0BgKiq0DGdvOKUqJn-FdgbF0LSVykwf__FY8/edit?pli=1)
- [0002, UI Modeling, 2024-01-08](https://docs.google.com/document/d/1tP63Ob4XPBGlASPHVf1GnNb3hlIkybmWBqJDA61u3J0/edit)
- https://github.com/icefoganalytics/internal-data-portal/issues/38

# Context

Adds Subscriptions panel excluding the ability to email users.

## Dataset Access Page - /datasets/:slug/access

### Front-end routes

Only the page route, though one could argue that the nearby tabs should be included. Note of the components on this page redirect away from it, they just open modals

### Back-end routes

* /api/datasets/:id - GET or maybe /datasets/:slug_or_id
* /api/access-grants?where[data_set_id=... - Get all access grants for a given dataset
* /api/access-grants - POST, create a access grant
* /api/access-grants/:id - PATCH, update an access grant
* /api/access-grants/:id - DELETE, delete an access grant
* /api/access-requests?where[data_set_id]=... - GET all access requests for a given dataset
* /api/access-requests - POST, create new access request
* /api/access-requests/:id/revoke - POST, revoke and access request
* /api/access-requests/:id/approve - POST, approve an access request
* /api/access-requests/:id/deny - POST, deny and access request
* /api/email-templates/:slug/new?data_set_id=.. - virtual model, for now, generates the prefilled content of notify users email
* /api/emails - POST, enqueue a new email job

### Components

* `BaseLayout.vue` - see previous definition
* `DatasetLayout.vue` - see previous definition
* `DatasetAccessPage.vue` - only available to higher rank users
    * `SubscriptionsCard.vue`
        * `DenyAccessDialog.vue` - see earlier implementation
            * AccessRequest#denier_id, 
            * AccessRequest#denied_at
            * AccessRequest#denial_reason
        * `AccessRequestManageTable.vue`
            * First Name - AccessRequest#requestor_id -> User#first_name
            * Last Name - AccessRequest#requestor_id -> User#last_name
            * Department - AccessRequest#requestor_id -> User#department
            * Use - AccessRequest#project_description
            * Access Type - AccessRequest#access_type
            * Status - generated from AccessRequest#approved_at/denided_at
            * License - AccessRequest#data_set_id -> Dataset#subscription_license
            * (unlabeled) - actions column
                * `RevokeAccessButton.vue` - revoke an approved access request or subscription, updates AccessRequest#revoker_id, and AccessRequest#revoked_at
                * `ApproveAccessButton.vue` - approve an access request, updates AccessRequest#approver_id, and AccessRequest#approved_at
                * `DenyAccessButton.vue` - opens deny request dialog

### Source Wireframe

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/3df034b2-899f-43b1-8fb2-5a17b057dd80)

# Implementation

Add a new table for subscriptions (Access Requests).
Add a QA Scenario for generating access requests.
Add stateful endpoints for access request approve/deny/revoke.

# Screenshots

![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/abd35dac-cf11-4418-a8df-85e842bd3527)

Revoke access dialog
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/af003d67-5c7d-4ffc-a2c0-513a32ac99bc)

Approve access dialog
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/4a7aa68d-a6ca-4f7e-988c-3804372ebe56)

Deny access dialog
![image](https://github.com/icefoganalytics/internal-data-portal/assets/23045206/c62b3a2c-764f-4009-8099-80709ee3a81b)

# Testing Instructions

1. Run the test suite via `dev test` (or `dev test_api`)
2. Boot the app via `dev up`
3. Log in to the app at http://localhost:8080
4. From the /dasboard, create a dataset as needed.
5. Use the QA Scenarios -> Apply Rand Access Grants or apply some to a dataset manually.
6. Use QA Scenarios -> Add Random Subscriptions (Access Requests) to add some access requests to all datasets.
7. Click the View Datasets button to get to the datasets page.
8. Click on a dataset to go the the - not built - description page for the dataset.
9. Click on the Access tab and scroll to the bottom to see the "Subscriptions" a.k.a. access requests.
10. Check that you can approve/deny/revoke requests as appropriate.
> Note that revoked and denied requests no longer appear in the table.
